packages/rif-id-core/src/operations/authentication.ts

```diff
@@ -0,0 +1,56 @@
+import { Dispatch } from '@reduxjs/toolkit'
+import { Agent, Credential, Message } from 'daf-core'
+import { Callback, callbackify } from './util'
+import axios from 'axios'
+import { ActionSignW3cVc } from 'daf-w3c'
+import { addServiceToken } from '../reducers/authentication'
+
+export const serviceLoginFactory = (agent: Agent) => (serviceUrl: string, serviceDid: string, did: string, cb?: Callback<string>) => (dispatch: Dispatch) => callbackify(
+  async () => {
+    const makeLoginCredentialPayload = (challenge: string) => {
+      if (!challenge) {
+        throw new Error('Server did not return any challenge to login')
+      }
+
+      return {
+        '@context': ['https://www.w3.org/2018/credentials/v1'],
+        type: ['VerifiableCredential'],
+        issuer: did,
+        credentialSubject: {
+          id: serviceDid,
+          claims: [{ claimType: 'challenge', claimValue: challenge }]
+        }
+      }
+    }
+
+    const verifyServiceDid = (message: Message): Credential => {
+      const credential = message.credentials[0]
+
+      if (credential.issuer.did !== serviceDid) {
+        throw new Error('The issuer of the auth credential is not the expected did')
+      }
+
+      return credential
+    }
+
+    return axios.post(`${serviceUrl}/request-auth`, { did })
+      .then(res => res.status === 200 && res.data)
+      .then(data => data.challenge)
+      .then(makeLoginCredentialPayload)
```

Comment:
I partially agree on that. It's true that it will become simpler to create just a signed JWT instead of a VC, but take in account that `daf-w3c` already has an action handler to build those VCs, and if we want to sign a JWT we should build an action handler to do that within `daf-did-jwt` because it is not created yet, or use directly `did-jwt` package with no agent involved, we just need to obtain the `signer` from it.
I think is cleaner to create a `daf-did-jwt` action handler inside `rif-id-daf` package

Reply:
In another hand, if we change the type of the JWT we should change the auth package as well, it is expecting the challenge inside a W3C VC
https://github.com/rsksmart/rif-identity-services/blob/cc4b183b9dc3d60e958be1da87afeafb87aba9ad/rif-id-jwt-auth/src/index.js#L76-L78

Reply:
> we should build an action handler to do that within `daf-did-jwt`

that's a bummer...

> I think is cleaner to create a daf-did-jwt action handler inside rif-id-daf package

I would rise an issue in uPort project to see if they are interested in merging this, what do you think?

> we should change the auth package as well

IMHO it worths it

---

Anyway, before taking any desicion let's think a little bit better on this interface. I guess a good extension in the future will include some credentials in this step

1. User requests access
2. Services sends a token and a selective disclosure request
3. User fills request with declarative details and/or presentations + challange
4. Service grants access and gives credential to user, probably including metadata taken from the request response.

Reply:
Based on what we already agreed, let's do the following:
1. User requests access
2. Service sends a challenge
3. Users signs a VP with the received challenge in the VP metadata. The user should include the needed VC inside that VP, if the service does not require any VC, it should send an empty array
4. Service validates the challenge and performs the needed business validations with the received VCs
5. If all the validations pass, the service creates a VC and sends it JWT representation to the user
6. User uses the received VC to authenticate each request to the service

Remarks:
- Why the user needs to include VCs in the authentication process? Because the service may need to verify some business logic before granting access, for instance, that the user is older than 18 years old
- Why the user will authenticate with the received VC and not with a ad hoc VP for each request? Because there is not added value by doing that and it needs an extra signing step with no sense

Reply:
Created a new issue in [rif-identity-services](https://github.com/rsksmart/rif-identity-services/issues/34 )